ihmc-perception/src/main/resources/us/ihmc/perception/gpuHeightMap/RapidHeightMapExtractor.cu

```diff
@@ -0,0 +1,580 @@
+                heightSum += globalMap[index] / params[HEIGHT_SCALING_FACTOR] - params[HEIGHT_OFFSET];
+                count++;
+            }
+        }
+    }
+    float heightAverage = heightSum / (float) count;
+    return heightAverage;
+}
+
+__device__ float get_spatial_stddev(int xIndex, int yIndex, float average, unsigned short *globalMap, float *params)
+{
+    float totalDeviation = 0.0f;
+    int count = 0;
+    int globalCellsPerAxis = (int)params[GLOBAL_CELLS_PER_AXIS];
+
+    for (int i = -1; i < 2; i++)
+    {
+        for (int j = -1; j < 2; j++)
+        {
+            int nxIndex = xIndex + i;
+            int nyIndex = yIndex + j;
+
+            if (nxIndex >= 0 && nxIndex < globalCellsPerAxis && nyIndex >= 0 && nyIndex < globalCellsPerAxis)
+            {
+                int index = nyIndex * globalCellsPerAxis + nxIndex;
+                float height = globalMap[index] / params[HEIGHT_SCALING_FACTOR] - params[HEIGHT_OFFSET];
+                totalDeviation += (height - average) * (height - average);
+                count++;
+            }
+        }
+    }
+float heightStddev = sqrt(totalDeviation / (float)count);
+   return heightStddev;
+
+}
+
+__device__ float get_spatial_filtered_height(int xIndex, int yIndex, float height, unsigned short *globalMap, float *params)
+{
+    float averageHeightZ = get_spatial_average(xIndex, yIndex, globalMap, params);
+    float heightStddev = get_spatial_stddev(xIndex, yIndex, averageHeightZ, globalMap, params);
+    float finalHeight = height;
+
+    if (fabs(finalHeight - averageHeightZ) < 0.5f * heightStddev)
+    {
+        //      finalHeight = averageHeightZ * params[SPATIAL_ALPHA] + finalHeight * (1.0f - params[SPATIAL_ALPHA]);
+    }
+    else
+    {
+        finalHeight = averageHeightZ * params[SPATIAL_ALPHA] * 0.0001f + finalHeight * (1.0f - params[SPATIAL_ALPHA] * 0.0001f);
+    }
+
+    return finalHeight;
+}
+
+// Compute grid cell center coordinates (cellCenterInZUp) in the Z-Up frame based on thread indices.
+// Transform the grid cell to the sensor frame using the transformation matrix (zUpToSensorFrameTf).
+// Perform projection (spherical or perspective) to map the grid cell to image indices.
+// Iterate over a search window in the depth image to find points within the cell bounds.
+// Back-project these points to the 3D space and transform them back to the Z-Up frame.
+// Compute the average height for points within the grid cell while filtering outliers.
+
+
+extern "C" __global__ void preprocessImageKernel(unsigned short *in, size_t pitchIn, unsigned short *out, size_t pitchOut)
+{
+    // Thread indices for output image
+    int xIndex = blockIdx.x * blockDim.x + threadIdx.x; // Output dimension x (new cols)
+    int yIndex = blockIdx.y * blockDim.y + threadIdx.y; // Output dimension y (new rows)
+
+    // Ensure within bounds of the output image (1280 rows and 720 columns)
+    if (xIndex >= 1280 || yIndex >= 720) {
```

Comment:
I think this hard coded resolution needs to be passed in. Maybe add a TODO somewhere about it if we don't cover it in this PR